File: solutions/R/trees_and_ensembles_ex2.R
# =============================================================================
# Chapter 8, Exercise 2: Random Forest vs XGBoost Tuning Challenge
# 1. Split data 80/20. 2. Tune RF and XGBoost with 5-fold CV.
# 3. Select best hyperparameters. 4. Evaluate on test set.
# 5. Create variable importance plots.
# =============================================================================

library(tidyverse)
library(tidymodels)

# --- Simulate the readmission dataset (same as chapter) ---
set.seed(42)
n <- 1000

readmit_data <- tibble(
  age = rnorm(n, 68, 12),
  length_of_stay = rpois(n, 5) + 1,
  num_comorbidities = rpois(n, 3),
  prior_admissions = rpois(n, 1),
  discharge_hemoglobin = rnorm(n, 11, 2),
  discharge_creatinine = rlnorm(n, 0.2, 0.5),
  has_diabetes = rbinom(n, 1, 0.35),
  has_chf = rbinom(n, 1, 0.25),
  readmitted = factor(
    rbinom(n, 1, plogis(-3 + 0.02 * (rnorm(n, 68, 12) - 68) +
                          0.15 * rpois(n, 1) +
                          0.1 * rpois(n, 3) +
                          0.3 * rbinom(n, 1, 0.25) -
                          0.1 * rnorm(n, 11, 2))),
    labels = c("No", "Yes")
  )
)

# --- Part 1: Train/test split ---
set.seed(42)
split <- initial_split(readmit_data, prop = 0.8, strata = readmitted)
train <- training(split)
test  <- testing(split)

cat("Training set size:", nrow(train), "\n")
cat("Test set size:", nrow(test), "\n")

# 5-fold CV on training set
folds <- vfold_cv(train, v = 5, strata = readmitted)

# Recipe (shared)
base_recipe <- recipe(readmitted ~ ., data = train)

# --- Part 2a: Tune Random Forest ---
rf_spec <- rand_forest(
  trees = 500,
  mtry = tune(),
  min_n = tune()
) %>%
  set_engine("ranger", importance = "impurity") %>%
  set_mode("classification")

rf_wf <- workflow() %>%
  add_model(rf_spec) %>%
  add_recipe(base_recipe)

rf_grid <- grid_regular(
  mtry(range = c(2, 7)),
  min_n(range = c(5, 30)),
  levels = 5
)

set.seed(42)
rf_tune <- tune_grid(rf_wf, resamples = folds, grid = rf_grid,
                     metrics = metric_set(roc_auc))

cat("\n--- Random Forest Tuning Results (Top 5) ---\n")
print(show_best(rf_tune, metric = "roc_auc", n = 5))

best_rf <- select_best(rf_tune, metric = "roc_auc")
cat("\nBest RF params - mtry:", best_rf$mtry, "min_n:", best_rf$min_n, "\n")

# --- Part 2b: Tune XGBoost ---
xgb_spec <- boost_tree(
  trees = 500,
  tree_depth = tune(),
  learn_rate = tune(),
  min_n = tune()
) %>%
  set_engine("xgboost") %>%
  set_mode("classification")

xgb_wf <- workflow() %>%
  add_model(xgb_spec) %>%
  add_recipe(base_recipe)

xgb_grid <- grid_regular(
  tree_depth(range = c(2, 6)),
  learn_rate(range = c(-3, -1)),   # log10 scale: 0.001 to 0.1
  min_n(range = c(5, 20)),
  levels = 4
)

set.seed(42)
xgb_tune <- tune_grid(xgb_wf, resamples = folds, grid = xgb_grid,
                      metrics = metric_set(roc_auc))

cat("\n--- XGBoost Tuning Results (Top 5) ---\n")
print(show_best(xgb_tune, metric = "roc_auc", n = 5))

best_xgb <- select_best(xgb_tune, metric = "roc_auc")
cat("\nBest XGB params - depth:", best_xgb$tree_depth,
    "learn_rate:", best_xgb$learn_rate,
    "min_n:", best_xgb$min_n, "\n")

# --- Part 3: Finalize and fit on full training set ---
final_rf_wf <- finalize_workflow(rf_wf, best_rf)
final_xgb_wf <- finalize_workflow(xgb_wf, best_xgb)

rf_final_fit <- fit(final_rf_wf, data = train)
xgb_final_fit <- fit(final_xgb_wf, data = train)

# --- Part 4: Evaluate on test set ---
rf_test_pred  <- predict(rf_final_fit, test, type = "prob") %>%
  bind_cols(test %>% select(readmitted))
xgb_test_pred <- predict(xgb_final_fit, test, type = "prob") %>%
  bind_cols(test %>% select(readmitted))

rf_auc  <- roc_auc(rf_test_pred, truth = readmitted, .pred_Yes)
xgb_auc <- roc_auc(xgb_test_pred, truth = readmitted, .pred_Yes)

cat("\n=== Test Set Performance ===\n")
cat("Random Forest AUC:", rf_auc$.estimate, "\n")
cat("XGBoost AUC:      ", xgb_auc$.estimate, "\n")

if (xgb_auc$.estimate > rf_auc$.estimate) {
  cat("\nXGBoost performs better on the test set.\n")
} else {
  cat("\nRandom Forest performs better on the test set.\n")
}

# --- Part 5: Variable importance plots ---
# NOTE: the `vip` package was archived from CRAN and is no longer installable,
# so we read the importance scores straight out of the fitted engine objects
# and plot them ourselves. This is a few more lines but has no dependency and
# makes it obvious which importance measure is being shown.

importance_plot <- function(scores, title) {
  tibble(variable = names(scores), importance = as.numeric(scores)) |>
    slice_max(importance, n = 8) |>
    ggplot(aes(x = importance, y = reorder(variable, importance))) +
    geom_col(fill = "steelblue") +
    labs(x = "Importance", y = NULL, title = title) +
    theme_minimal()
}

# Random Forest: ranger stores impurity importance because the spec above set
# `importance = "impurity"`.
rf_scores <- rf_final_fit |>
  extract_fit_engine() |>
  (\(x) x$variable.importance)()
print(importance_plot(rf_scores, "Random Forest variable importance (impurity)"))

# XGBoost: xgb.importance() returns a data frame, so reshape it to a named vector.
xgb_imp <- xgboost::xgb.importance(model = extract_fit_engine(xgb_final_fit))
xgb_scores <- setNames(xgb_imp$Gain, xgb_imp$Feature)
print(importance_plot(xgb_scores, "XGBoost variable importance (gain)"))

cat("\nBoth models should generally agree on the most important features,\n")
cat("though rankings may differ. Continuous variables with more possible\n")
cat("split points (e.g., age, creatinine) often rank higher in tree-based\n")
cat("importance measures than binary variables (e.g., has_chf).\n")

File: solutions/python/trees_and_ensembles_ex2.py
# =============================================================================
# Chapter 8, Exercise 2: Random Forest vs XGBoost Tuning Challenge
# 1. Split data 80/20. 2. Tune RF and XGBoost with 5-fold CV.
# 3. Select best hyperparameters. 4. Evaluate on test set.
# 5. Create variable importance plots.
# =============================================================================

import numpy as np
import pandas as pd
from sklearn.model_selection import (train_test_split, RandomizedSearchCV,
                                      StratifiedKFold)
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import roc_auc_score
import xgboost as xgb
import matplotlib.pyplot as plt

# --- Simulate the readmission dataset (same as chapter) ---
np.random.seed(42)
n = 1000

df = pd.DataFrame({
    'age': np.random.normal(68, 12, n),
    'length_of_stay': np.random.poisson(5, n) + 1,
    'num_comorbidities': np.random.poisson(3, n),
    'prior_admissions': np.random.poisson(1, n),
    'discharge_hemoglobin': np.random.normal(11, 2, n),
    'discharge_creatinine': np.random.lognormal(0.2, 0.5, n),
    'has_diabetes': np.random.binomial(1, 0.35, n),
    'has_chf': np.random.binomial(1, 0.25, n)
})
y = np.random.binomial(1, 0.18, n)

feature_names = list(df.columns)

# --- Part 1: Train/test split ---
X_train, X_test, y_train, y_test = train_test_split(
    df, y, test_size=0.2, stratify=y, random_state=42
)
print(f"Training set: {X_train.shape[0]} | Test set: {X_test.shape[0]}")

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# --- Part 2a: Tune Random Forest ---
rf_param_dist = {
    'n_estimators': [200, 500],
    'max_features': ['sqrt', 'log2', 3, 5, 7],
    'min_samples_leaf': [5, 10, 15, 20, 30],
    'max_depth': [5, 10, 15, 20, None]
}

rf = RandomForestClassifier(random_state=42, n_jobs=-1)

rf_search = RandomizedSearchCV(
    rf, rf_param_dist, n_iter=30, cv=cv, scoring='roc_auc',
    random_state=42, n_jobs=-1, verbose=0
)
rf_search.fit(X_train, y_train)

print(f"\n--- Random Forest Tuning ---")
print(f"Best CV AUC: {rf_search.best_score_:.3f}")
print(f"Best params: {rf_search.best_params_}")

# --- Part 2b: Tune XGBoost ---
xgb_param_dist = {
    'n_estimators': [200, 500, 1000],
    'learning_rate': [0.01, 0.05, 0.1],
    'max_depth': [2, 3, 4, 5, 6],
    'subsample': [0.7, 0.8, 0.9, 1.0],
    'colsample_bytree': [0.7, 0.8, 0.9, 1.0],
    'min_child_weight': [1, 5, 10]
}

xgb_model = xgb.XGBClassifier(
    random_state=42, use_label_encoder=False, eval_metric='logloss'
)

xgb_search = RandomizedSearchCV(
    xgb_model, xgb_param_dist, n_iter=30, cv=cv, scoring='roc_auc',
    random_state=42, n_jobs=-1, verbose=0
)
xgb_search.fit(X_train, y_train)

print(f"\n--- XGBoost Tuning ---")
print(f"Best CV AUC: {xgb_search.best_score_:.3f}")
print(f"Best params: {xgb_search.best_params_}")

# --- Part 4: Evaluate on test set ---
rf_best = rf_search.best_estimator_
xgb_best = xgb_search.best_estimator_

rf_test_probs = rf_best.predict_proba(X_test)[:, 1]
xgb_test_probs = xgb_best.predict_proba(X_test)[:, 1]

rf_test_auc = roc_auc_score(y_test, rf_test_probs)
xgb_test_auc = roc_auc_score(y_test, xgb_test_probs)

print(f"\n=== Test Set Performance ===")
print(f"Random Forest AUC: {rf_test_auc:.3f}")
print(f"XGBoost AUC:       {xgb_test_auc:.3f}")

if xgb_test_auc > rf_test_auc:
    print("\nXGBoost performs better on the test set.")
else:
    print("\nRandom Forest performs better on the test set.")

# --- Part 5: Variable importance plots ---
fig, axes = plt.subplots(1, 2, figsize=(14, 5))

# Random Forest importance
rf_imp = pd.Series(rf_best.feature_importances_, index=feature_names)
rf_imp = rf_imp.sort_values(ascending=True)
rf_imp.plot(kind='barh', ax=axes[0], color='#2E86AB')
axes[0].set_xlabel("Feature Importance (Gini)")
axes[0].set_title("Random Forest Variable Importance")

# XGBoost importance
xgb_imp = pd.Series(xgb_best.feature_importances_, index=feature_names)
xgb_imp = xgb_imp.sort_values(ascending=True)
xgb_imp.plot(kind='barh', ax=axes[1], color='#D55E00')
axes[1].set_xlabel("Feature Importance (Gain)")
axes[1].set_title("XGBoost Variable Importance")

plt.tight_layout()
plt.show()

# Compare top features
print("\nRF top 3 features:", rf_imp.sort_values(ascending=False).head(3).index.tolist())
print("XGB top 3 features:", xgb_imp.sort_values(ascending=False).head(3).index.tolist())
print("\nBoth models should generally agree on the most important features,")
print("though rankings may differ due to how each algorithm uses features.")
